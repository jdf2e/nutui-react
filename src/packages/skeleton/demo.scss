.content {
  margin-bottom: 20px;
  .nut-switch {
    display: flex;
    margin: 0 16px 8px 0;
  }
  .container {
    display: flex;
    .right-content {
      margin-left: 19px;
      font-family: PingFangSC;
      display: flex;
      flex-direction: column;
      .title {
        font-size: 14px;
        color: rgba(51, 51, 51, 1);
      }
      .desc {
        margin-top: 10px;
        font-size: 13px;
        color: rgba(154, 155, 157, 1);
      }
    }
  }
}
.pic-compose {
  display: flex;
  justify-content: space-between;
  .item {
    width: 47%;
  }
}
