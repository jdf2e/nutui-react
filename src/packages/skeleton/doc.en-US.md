# Skeleton



Filling gray bitmap in the area to be loaded on the page is essentially the transition effect in the process of interface loading.

## Import

```tsx
import { Skeleton } from '@nutui/nutui-react';
```

## Code demonstration

### Basic usage

:::demo

<CodeBlock src='h5/demo1.tsx'></CodeBlock>

:::

### Incoming multiline

:::demo

<CodeBlock src='h5/demo2.tsx'></CodeBlock>

:::

### Show Faces

:::demo

<CodeBlock src='h5/demo3.tsx'></CodeBlock>

:::

### Title paragraph fillet style

:::demo

<CodeBlock src='h5/demo4.tsx'></CodeBlock>

:::

### Display subcomponents

:::demo

<CodeBlock src='h5/demo5.tsx'></CodeBlock>

:::

## Skeleton

### Props

| Property | Description | Type | Default |
| --- | --- | --- | --- |
| visible | Whether to display skeleton screen | `boolean` | `false` |
| animated | Whether to turn on skeleton screen animation | `boolean` | `false` |
| avatar | Show avatar | `boolean` | `false` |
| avatarShape | Avatar shape: square / round | `string` | `round` |
| avatarSize | Avatar size | `string` | `50px` |
| rows | Set the number of paragraph lines | `number` | `1` |
| title | Show paragraph titles | `boolean` | `false` |

## Theming

### CSS Variables

The component provides the following CSS variables, which can be used to customize styles. Please refer to [ConfigProvider component](#/en-US/component/configprovider).

| Name | Description | Default |
| --- | --- | --- |
| \--nutui-skeleton-background | background | `rgb(239, 239, 239)` |
| \--nutui-skeleton-line-width | line width | `100%` |
| \--nutui-skeleton-line-height | line height | `15px` |
| \--nutui-skeleton-line-border-radius | line borderRadius | `0` |