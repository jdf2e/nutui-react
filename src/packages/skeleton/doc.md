#  Skeleton组件

### 介绍

在页面上待加载区域填充灰色的占位图，本质上是界面加载过程中的过渡效果。

### 安装
``` ts
import { Skeleton } from '@nutui/nutui-react';
```


## 代码演示

### 基础用法

:::demo
```tsx
import React from "react";
import { Skeleton } from '@nutui/nutui-react';

const App = () => {
  return (
    <>
      <Skeleton width={250} height={15} animated></Skeleton>
    </>
  )
}
export default App;
```

### 传入多行

```tsx
import React from "react";
import { Skeleton } from '@nutui/nutui-react';

const App = () => {
  return (
    <>
      <Skeleton width={250} height={15} row={3} title animated></Skeleton>
    </>
  )
}
export default App;
```

### 显示头像

```tsx
import React from "react";
import { Skeleton } from '@nutui/nutui-react';

const App = () => {
  return (
    <>
      <Skeleton width={250} height={15} row={3} title animated avatar></Skeleton>
    </>
  )
}
export default App;
```

### 标题段落圆角风格

```tsx
import React from "react";
import { Skeleton } from '@nutui/nutui-react';

const App = () => {
  return (
    <>
      <Skeleton width={250} height={15} animated round></Skeleton>
    </>
  )
}
export default App;
```

### 显示子组件

```tsx
import React, { useState } from 'react'
import { Skeleton, Switch, Avatar } from '@nutui/nutui-react';

const NutSwitchStyle = {
  display: 'flex',
  margin: '0 16px 8px 0'
}

const RightContentStyle = {
  marginLeft: '19px',
  fontFamily: 'PingFangSC',
  display: 'flex',
  flexDirection: 'column'
}

const TitleStyle = {
  fontSize: '14px',
  color: 'rgba(51, 51, 51, 1)'
}

const DescStyle = {
  marginTop: '10px',
  fontSize: '13px',
  color: 'rgba(154, 155, 157, 1)'
}

const App = () => {
  const [checked, setChecked] = useState(false)
  const changeStatus = (value: boolean, event: React.MouseEvent<Element, MouseEvent>) => {
    console.log(`触发了change事件，开关状态：${value}`)
    setChecked(value)
  }
  return (
    <>
      <div className="content">
        <Switch size="15px" change={(value, event) => changeStatus(value, event)} className={NutSwitchStyle}></Switch>
        <Skeleton width={250} height={15} title animated avatar row={3} loading={checked}>
          <div className="container" style={{ display: 'flex' }}>
            <Avatar
              size="50"
              icon="https://img14.360buyimg.com/imagetools/jfs/t1/167902/2/8762/791358/603742d7E9b4275e3/e09d8f9a8bf4c0ef.png"
            />
            <div className="right-content" className={RightContentStyle}>
              <span className="title" className={TitleStyle}>NutUI-React</span>
              <div className="desc" className={DescStyle}>
                一套京东风格的轻量级移动端React组件库，提供丰富的基础组件和业务组件，帮助开发者快速搭建移动应用。
              </div>
            </div>
          </div>
        </Skeleton>
      </div>
    </>
  )
}
export default App;
```




## API

### Prop  

| 字段       | 说明                                             | 类型    | 默认值    |
|------------|-------------------------------------------------|---------|----------|
| loading    | 是否显示骨架屏                                    | Boolean | `false`    | 
| width       | 每行宽度                                        | Number  | 100 |
| height      | 每行高度                                        | Number  | 100   |
| animated    | 是否开启骨架屏动画                                | Boolean  | `false`  |
| avatar      | 是否显示头像                                     | Boolean | `false`   |
| avatarShape      | 头像形状：正方形/圆形                        | String | `round`   |
| avatarSize       | 头像大小                                   | String | `50px`    |
| round  | 标题/段落是否采用圆角风格                                | Boolean | `false`  |
| row    | 设置段落行数                                           | Number | 1       |
| title  | 是否显示段落标题                                        | Boolean | `false`   |
