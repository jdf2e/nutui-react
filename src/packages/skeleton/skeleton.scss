@import '../avatar/avatar.scss';

.nut-skeleton {
  display: inline-block;
  position: relative;
  overflow: hidden;
  vertical-align: middle;
  width: 100%;

  &-content {
    display: flex;
    flex-direction: row;

    &-avatar {
      margin-right: 20px;
      background: $skeleton-background !important;
    }

    &-block {
      width: $skeleton-line-width;
      height: $skeleton-line-height;
      background: $skeleton-background;
      margin-top: 10px;
      border-radius: $skeleton-line-border-radius;
    }

    &-line {
      width: 100%;
      display: flex;
      flex-direction: column;
      .nut-skeleton-content-block:last-child {
        width: 70%;
      }
    }

    &-title {
      width: 30%;
      height: $skeleton-line-height;
      background: $skeleton-background;
    }

    &-block-last-child {
      width: 55%;
    }
  }

  &-animation {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1;
    // background: linear-gradient(
    //   90deg,
    //   hsla(0, 0%, 100%, 0),
    //   hsla(0, 0%, 100%, 0.5) 50%,
    //   hsla(0, 0%, 100%, 0) 80%
    // );
    animation: nut-skeleton 2s ease-in-out 0s infinite;
  }

  @keyframes nut-skeleton {
    0% {
      background-position-x: -500px;
    }

    to {
      background-position-x: calc(500px + 100%);
    }
  }
}

[dir='rtl'] .nut-skeleton,
.nut-rtl .nut-skeleton {
  &-content {
    .nut-avatar {
      margin-right: 0;
      margin-left: 20px;
    }
  }
  &-animation {
    left: auto;
    right: 0;
    // background: linear-gradient(
    //   -90deg,
    //   hsla(0, 0%, 100%, 0),
    //   hsla(0, 0%, 100%, 0.5) 50%,
    //   hsla(0, 0%, 100%, 0) 80%
    // );
    animation: nut-skeleton-rtl 2s ease-in-out 0s infinite;
  }

  @keyframes nut-skeleton-rtl {
    0% {
      background-position-x: calc(500px + 100%);
    }

    to {
      background-position-x: -500px;
    }
  }
}
