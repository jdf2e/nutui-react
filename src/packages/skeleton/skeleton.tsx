import React, { FunctionComponent } from 'react'
import classNames from 'classnames'
import Avatar from '@/packages/avatar'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'

type avatarShape = 'round' | 'square'
export interface SkeletonProps extends BasicComponent {
  animated: boolean
  rows: number
  title: boolean
  avatar: boolean
  avatarSize: string
  round: boolean
  visible: boolean
  avatarShape: avatarShape
}
const defaultProps = {
  ...ComponentDefaults,
  rows: 1,
  animated: false,
  title: false,
  avatar: false,
  round: false,
  avatarSize: '50px',
  visible: false,
  avatarShape: 'round',
} as SkeletonProps
export const Skeleton: FunctionComponent<Partial<SkeletonProps>> = (props) => {
  const {
    className,
    animated,
    rows,
    title,
    avatar,
    avatarSize,
    round,
    visible,
    children,
    avatarShape,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }

  const classPrefix = 'nut-skeleton'
  const classes = classNames(className, classPrefix)
  const blockClass = classNames({
    block: true,
    'block--round': round,
  })
  const avatarClass = classNames({
    avatar: true,
    [`avatar--${avatarShape}`]: avatarShape,
  })

  const repeatLines = (num: number) => {
    return Array.from({ length: num }, (v, i) => i)
  }

  const getStyle = () => {
    if (avatarSize) {
      return {
        width: avatarSize,
        height: avatarSize,
      }
    }
    return {
      width: '50px',
      height: '50px',
    }
  }

  return (
    <>
      {visible ? (
        <div>{children}</div>
      ) : (
        <div className={classes} {...rest}>
          {animated && <div className={`${classPrefix}__animation`} />}
          <div className={`${classPrefix}__content`}>
            {avatar && (
              <Avatar
                className={avatarClass}
                background="rgb(239, 239, 239)"
                shape={avatarShape}
                style={getStyle()}
                icon="null"
              />
            )}
            {rows === 1 ? (
              <div className={blockClass} />
            ) : (
              <div className={`${classPrefix}__content-line`}>
                {title && <div className={`${classPrefix}__title`} />}
                {repeatLines(rows).map((item, index) => {
                  return <div className={blockClass} key={index} />
                })}
              </div>
            )}
          </div>
        </div>
      )}
    </>
  )
}

Skeleton.defaultProps = defaultProps
Skeleton.displayName = 'NutSkeleton'
