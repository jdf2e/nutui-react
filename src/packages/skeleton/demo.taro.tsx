import React, { useState } from 'react'
import Taro from '@tarojs/taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import {
  Skeleton,
  Switch,
  Avatar,
  Cell,
  ConfigProvider,
} from '@/packages/nutui.react.taro'
import '@/packages/skeleton/demo.scss'
import Header from '@/sites/components/header'

interface T {
  '3b02fdee': string
  '84aa6bce': string
  ea3bc18a: string
  '02a53df5': string
  '0a001122': string
  a4ed11b5: string
  '07d62d5c': string
  '652c6725': string
}
const SkeletonDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      '3b02fdee': '触发了change事件，开关状态：',
      '84aa6bce': '基础用法',
      ea3bc18a: '传入多行',
      '02a53df5': '显示头像',
      '0a001122': '标题段落圆角风格',
      a4ed11b5: '图片组合',
      '07d62d5c': '显示子组件',
      '652c6725':
        '一套京东风格的轻量级移动端React组件库，提供丰富的基础组件和业务组件，帮助开发者快速搭建移动应用。',
    },
    'zh-TW': {
      '3b02fdee': '觸發了change事件，開關狀態：',
      '84aa6bce': '基礎用法',
      ea3bc18a: '傳入多行',
      '02a53df5': '顯示頭像',
      '0a001122': '標題段落圓角風格',
      a4ed11b5: '图片组合',
      '07d62d5c': '圖片組合',
      '652c6725':
        '一套京東風格的輕量級移動端React組件庫，提供豐富的基礎組件和業務組件，幫助開發者快速搭建移動應用。',
    },
    'en-US': {
      '3b02fdee': 'The change event is triggered, the switch state:',
      '84aa6bce': 'Basic usage',
      ea3bc18a: 'Pass in multiple lines',
      '02a53df5': 'show avatar',
      '0a001122': 'Heading Paragraph Rounded Corner Style',
      a4ed11b5: 'picture combination',
      '07d62d5c': 'show subcomponents',
      '652c6725':
        'A set of JD-style lightweight mobile React component library, providing rich basic components and business components to help developers quickly build mobile applications.',
    },
  })

  const [checked, setChecked] = useState(false)
  const changeStatus = (
    value: boolean,
    event: React.MouseEvent<Element, MouseEvent>
  ) => {
    console.log(`${translated['3b02fdee']}${value}`)
    setChecked(value)
  }
  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>{translated['84aa6bce']}</h2>
        <Cell className="ske-cell-single">
          <Skeleton animated />
          <Skeleton />
        </Cell>

        <h2>{translated.ea3bc18a}</h2>
        <Cell className="ske-cell-double">
          <Skeleton rows={3} title animated />
        </Cell>

        <h2>{translated['02a53df5']}</h2>
        <Cell>
          <Skeleton rows={3} title animated avatar avatarSize="100px" />
        </Cell>

        <h2>{translated['0a001122']}</h2>
        <Cell className="ske-cell-single">
          <ConfigProvider
            theme={{
              nutuiSkeletonLineBorderRadius: '10px',
            }}
          >
            <Skeleton rows={3} animated />
          </ConfigProvider>
        </Cell>

        <h2>{translated.a4ed11b5}</h2>
        <Cell className="ske-cell-double">
          <div className="pic-compose">
            <Skeleton title animated rows={3} className="item" />
            <Skeleton title animated rows={3} className="item" />
          </div>
        </Cell>

        <h2>{translated['07d62d5c']}</h2>
        <Cell>
          <div className="content" style={{ width: '100%' }}>
            <Switch onChange={(value, event) => changeStatus(value, event)} />
            <Skeleton title animated avatar rows={3} visible={checked}>
              <div className="container">
                <Avatar
                  size="50"
                  icon={
                    <img
                      src="https://img14.360buyimg.com/imagetools/jfs/t1/167902/2/8762/791358/603742d7E9b4275e3/e09d8f9a8bf4c0ef.png"
                      alt=""
                    />
                  }
                />
                <div className="right-content">
                  <span className="title">NutUI-React</span>
                  <div className="description">{translated['652c6725']}</div>
                </div>
              </div>
            </Skeleton>
          </div>
        </Cell>
      </div>
    </>
  )
}

export default SkeletonDemo
