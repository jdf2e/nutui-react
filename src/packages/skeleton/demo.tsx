import React, { useState } from 'react'
import { Skeleton } from './skeleton'
import Switch from '@/packages/switch'
import Avatar from '@/packages/avatar'
import './demo.scss'

const SkeletonDemo = () => {
  const [checked, setChecked] = useState(false)
  const changeStatus = (value: boolean, event: React.MouseEvent<Element, MouseEvent>) => {
    console.log(`触发了change事件，开关状态：${value}`)
    setChecked(value)
  }
  return (
    <>
      <div className="demo">
        <h2>基础用法</h2>
        <Skeleton width="250px" height="15px" animated />
        <Skeleton width="250px" height="15px" />

        <h2>传入多行</h2>
        <Skeleton width="250px" height="15px" row={3} title animated />

        <h2>显示头像</h2>
        <Skeleton width="250px" height="15px" row={3} title animated avatar avatarSize="100px" />

        <h2>标题段落圆角风格</h2>
        <Skeleton width="250px" height="15px" animated round />

        <h2>图片组合</h2>
        <div className="pic-compose">
          <Skeleton width="250px" height="15px" title animated row={3} className="item" />
          <Skeleton width="250px" height="15px" title animated row={3} className="item" />
        </div>

        <h2>显示子组件</h2>
        <div className="content">
          <Switch size="15px" change={(value, event) => changeStatus(value, event)} />
          <Skeleton width="250px" height="15px" title animated avatar row={3} loading={checked}>
            <div className="container">
              <Avatar
                size="50"
                icon="https://img14.360buyimg.com/imagetools/jfs/t1/167902/2/8762/791358/603742d7E9b4275e3/e09d8f9a8bf4c0ef.png"
              />
              <div className="right-content">
                <span className="title">NutUI-React</span>
                <div className="desc">
                  一套京东风格的轻量级移动端React组件库，提供丰富的基础组件和业务组件，帮助开发者快速搭建移动应用。
                </div>
              </div>
            </div>
          </Skeleton>
        </div>
      </div>
    </>
  )
}

export default SkeletonDemo
