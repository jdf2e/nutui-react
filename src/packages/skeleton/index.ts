import { Skeleton } from './skeleton'

export default Skeleton
