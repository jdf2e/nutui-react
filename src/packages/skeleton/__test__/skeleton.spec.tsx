import * as React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'

import { Skeleton } from '../skeleton'

test('should render with row width array correctly', () => {
  const { container } = render(<Skeleton width="100px" />)
  expect(container.querySelector('.blockClass')).toHaveStyle({ width: '100px' })
})

test('should allow to disable animation', () => {
  const { container } = render(<Skeleton row={1} animated={false} />)
  expect(container.querySelector('.skeleton-animation')).toBeFalsy()
})

test('should change avatar size when using avatarSize prop', () => {
  const { container } = render(<Skeleton avatar avatarSize="20px" />)
  expect(container.querySelector('.avatarClass')).toHaveStyle({
    width: '20px',
    height: '20px',
  })
  expect(container).toMatchSnapshot()
})

test('should change avatar shape when using avatarShape prop', () => {
  const { container } = render(<Skeleton avatar avatarShape="square" />)
  expect(container.querySelector('.avatarClass')).toMatchSnapshot()
})

test('should be round when using round prop', () => {
  const { container } = render(<Skeleton title round avatar />)
  expect(container.querySelector('.avatarClass')).toBeTruthy()
})
