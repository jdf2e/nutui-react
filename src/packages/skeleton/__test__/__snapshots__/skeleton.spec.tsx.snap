// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should change avatar shape when using avatarShape prop 1`] = `
<div
  class="nut-avatar nut-avatar-normal nut-avatar-square nut-avatar nut-skeleton-content-avatar avatar-square"
  style="width: 50px; height: 50px; background-color: #eee; color: #666;"
/>
`;

exports[`should change avatar size when using avatarSize prop 1`] = `
<div>
  <div
    class="nut-skeleton"
  >
    <div
      class="nut-skeleton-content"
    >
      <div
        class="nut-avatar nut-avatar-normal nut-avatar-round nut-avatar-normal-round nut-avatar nut-skeleton-content-avatar avatar-round"
        style="width: 20px; height: 20px; background-color: #eee; color: #666;"
      />
      <div
        class="nut-skeleton-content-block"
      />
    </div>
  </div>
</div>
`;
